src/tiered/tiered_work.c

```diff
@@ -43,6 +43,31 @@ __wt_tiered_work_free(WT_SESSION_IMPL *session, WT_TIERED_WORK_UNIT *entry)
     __wt_free(session, entry);
 }
 
+/*
+ * __wt_tiered_remove_work --
+ *     Remove all work on the queue that applies to the given tiered handle.
+ */
+void
+__wt_tiered_remove_work(WT_SESSION_IMPL *session, WT_TIERED *tiered)
+{
+    WT_CONNECTION_IMPL *conn;
+    WT_TIERED_WORK_UNIT *entry, *entry_tmp;
+
+    conn = S2C(session);
+    __wt_spin_lock(session, &conn->tiered_lock);
+    TAILQ_FOREACH_SAFE(entry, &conn->tieredqh, q, entry_tmp)
+    {
+        /* Remove and free any entry for this tiered handle. */
+        if (entry->tiered == tiered) {
+            TAILQ_REMOVE(&conn->tieredqh, entry, q);
+            WT_STAT_CONN_INCR(session, tiered_work_units_removed);
+            __wt_tiered_work_free(session, entry);
```

Comment:
I have a feeling that just removing work units from the queue is not right... I might be wrong, but what if we have this scenario:
1. Create a table `table-0000001.wtobj`
2. Perform some inserts and flush_tier calls so that the active file is, let's say, `table-0000009.wtobj` and other object files are in the cache and shared storage.
3. Now we drop the table. `WT_TIERED_WORK_DROP_LOCAL`/`WT_TIERED_WORK_DROP_SHARED` work unit is added to the queue.
4. I suspect the tiered dhandle may still be in the cache.
5. While the work unit is in the queue we create a table with the same name and drop it. But at this time the active file is, for instance, `table-0000002.wtobj`. Will we reuse the dhandle from the previous table or create a new one here? Because if we reuse, we will discard the previous drop work unit. 

Would this leave us with unused files, that will cause a file exists error?

Reply:
I think the `WT_TIERED_WORK_DROP_LOCAL|SHARED` work may be poorly named. It is not related to `session->drop`. The `DROP_LOCAL` work unit is added at the time of a flush to remove the local copy of an object after some time, default 5 minutes. So the `DROP_LOCAL` is not added as part of the `drop` call but added as part of a `flush_tier`. The `DROP_SHARED` work unit is future intended to be used for garbage collection and is not used in the code. But I will think harder about the general problem of if some other thread could be racing by being in the middle of processing and then adding new units while we're dropping.

Reply:
I have opened WT-9527 for the naming issue. The scenario you mention does not happen because dropping the table doesn't add work units relating to the table to the queue. The whole point of this change is to remove anything relating to the dropped table from the queue.

Reply:
Thank you for the explanations!